components/cloud/gcp/src/gcs.rs

```diff
@@ -605,6 +630,83 @@ mod tests {
         assert_eq!(c1.bucket.prefix, c2.bucket.prefix);
     }
 
+    enum ThrottleReadState {
+        Spawning,
+        Emitting,
+    }
+    /// ThrottleRead throttles a `Read` -- make it emits 2 chars for each
+    /// `read` call. This is copy & paste from the implmentation from s3.rs.
+    #[pin_project::pin_project]
+    struct ThrottleRead<R> {
+        #[pin]
+        inner: R,
+        state: ThrottleReadState,
+    }
+    impl<R: AsyncRead> AsyncRead for ThrottleRead<R> {
+        fn poll_read(
+            self: std::pin::Pin<&mut Self>,
+            cx: &mut std::task::Context<'_>,
+            buf: &mut [u8],
+        ) -> Poll<io::Result<usize>> {
+            let this = self.project();
+            match this.state {
+                ThrottleReadState::Spawning => {
+                    *this.state = ThrottleReadState::Emitting;
+                    cx.waker().wake_by_ref();
+                    Poll::Pending
+                }
+                ThrottleReadState::Emitting => {
+                    *this.state = ThrottleReadState::Spawning;
+                    this.inner.poll_read(cx, &mut buf[..2])
+                }
+            }
+        }
+    }
+    impl<R> ThrottleRead<R> {
+        fn new(r: R) -> Self {
+            Self {
+                inner: r,
+                state: ThrottleReadState::Spawning,
+            }
+        }
+    }
+
+    const BENCH_READ_SIZE: usize = 128 * 1024;
+
+    // 88,852,361 ns/iter (+/- 11,536,292) (futures-util 0.3.15)
+    #[bench]
+    fn bench_read_to_end(b: &mut test::Bencher) {
+        b.iter(|| {
+            let mut v = [0; BENCH_READ_SIZE];
+
+            let mut r = ThrottleRead::new(Cursor::new(&mut v));
+            let mut dst = Vec::with_capacity(BENCH_READ_SIZE);
+            let rt = tokio::runtime::Builder::new_current_thread()
+                .build()
+                .unwrap();
```

Comment:
They are noises for a benchmark of `read_to_end`, they should be moved out the `b.iter()`.

```rust
let mut dst = Vec::with_capacity(BENCH_READ_SIZE);
let rt = tokio::runtime::Builder::new_current_thread()
    .build()
    .unwrap();
let mut v = [0; BENCH_READ_SIZE];
b.iter(|| {
    let mut r = ThrottleRead::new(Cursor::new(&mut v));
    dst.clear();
    rt.block_on(r.read_to_end(&mut dst)).unwrap();
    assert_eq!(dst.len(), BENCH_READ_SIZE)
})
```